File: bundles/handler-sample-app/src/main/java/io/wcm/samples/handler/business/navigation/NavigationManagerImpl.java
/*
 * #%L
 * wcm.io
 * %%
 * Copyright (C) 2014 wcm.io
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.wcm.samples.handler.business.navigation;

import io.wcm.handler.link.Link;
import io.wcm.handler.link.LinkHandler;
import io.wcm.samples.handler.util.SiteHelper;
import io.wcm.sling.models.annotations.AemObject;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import org.apache.commons.lang3.StringUtils;
import org.apache.sling.api.SlingHttpServletRequest;
import org.apache.sling.models.annotations.Model;
import org.apache.sling.models.annotations.injectorspecific.Self;

import com.day.cq.wcm.api.Page;
import com.day.cq.wcm.api.PageFilter;

/**
 * Generates navigation page items based on a page hierarchy for navigation controllers.
 * Only pages that are valid and not hidden with valid links are included.
 * Optionally, structure element pages without links can be included as well.
 */
@Model(adaptables = SlingHttpServletRequest.class, adapters = NavigationManager.class)
public class NavigationManagerImpl implements NavigationManager {

  @Self
  private SiteHelper siteHelper;
  @Self
  private LinkHandler linkHandler;
  @AemObject
  private Page currentPage;

  private static final String FOOTERNAV_RELATIVE_PATH = "/tools/navigation/footernav";

  /**
   * Generation main navigation links.
   * @param maxLevels Max. navigation hierarchy levels
   * @return Root page for navigation
   */
  @Override
  public NavigationPageItem getMainNavigation(final int maxLevels) {
    Page siteRootPage = siteHelper.getSiteRootPage();
    NavigationPageItem rootItem = createLinkableItem(siteRootPage);
    if (maxLevels > 0) {
      rootItem.setChildren(createChildItemsRecursively(siteRootPage, new ValidLinkableItemCreator(), 1, maxLevels));
    }
    return rootItem;
  }

  /**
   * Generation footer navigation links and navigation structure.
   * If a page '/tools/navigation/footerNav" exists this page is used for the root navigation hierarchy, using the first
   * level of pages as column structure and the next level as link entries.
   * If not present the main navigation sections are used as column structure, and the main navigation pages itself
   * including their children as link entries.
   * @return Root page for navigation
   */
  @Override
  public NavigationPageItem getFooterNavigation() {
    Page footerNavRoot = siteHelper.getRelativePage(FOOTERNAV_RELATIVE_PATH);
    if (footerNavRoot != null) {
      return getFooterNavigationSpecific(footerNavRoot);
    }
    else {
      return getFooterNavigationDerivedFromMainNav(siteHelper.getSiteRootPage());
    }
  }

  private NavigationPageItem getFooterNavigationSpecific(final Page footerNavRoot) {
    NavigationPageItem rootItem = createStructureItem(footerNavRoot);
    rootItem.setChildren(createChildItems(footerNavRoot, new ItemCreator() {
      @Override
      public NavigationPageItem create(final Page pPage) {
        NavigationPageItem columnItem = createStructureItem(pPage);
        columnItem.setChildren(createChildItems(pPage, new ValidLinkableItemCreator()));
        if (columnItem.getChildren().isEmpty()) {
          return null;
        }
        else {
          return columnItem;
        }
      }
    }));
    return rootItem;
  }

  private NavigationPageItem getFooterNavigationDerivedFromMainNav(final Page siteRootPage) {
    NavigationPageItem rootItem = createStructureItem(siteRootPage);
    rootItem.setChildren(createChildItems(siteRootPage, new ItemCreator() {
      @Override
      public NavigationPageItem create(final Page pPage) {
        NavigationPageItem columnItem = createStructureItem(pPage);
        List<NavigationPageItem> linkItems = new ArrayList<>();
        NavigationPageItem mainnavLinkItem = createLinkableItem(pPage);
        if (mainnavLinkItem.getLink().isValid()) {
          linkItems.add(mainnavLinkItem);
        }
        linkItems.addAll(createChildItems(pPage, new ValidLinkableItemCreator()));
        columnItem.setChildren(linkItems);
        if (columnItem.getChildren().isEmpty()) {
          return null;
        }
        else {
          return columnItem;
        }
      }
    }));
    return rootItem;
  }

  private NavigationPageItem createStructureItem(final Page page) {
    String title = getItemTitle(page);
    return new NavigationPageItem(title);

  }

  private NavigationPageItem createLinkableItem(final Page page) {
    String title = getItemTitle(page);
    boolean isCurrentPage = StringUtils.equals(page.getPath(), currentPage.getPath());
    Link link = linkHandler.get(page).build();
    return new NavigationPageItem(title, isCurrentPage, link);
  }

  private String getItemTitle(final Page page) {
    return StringUtils.defaultString(page.getNavigationTitle(), page.getTitle());
  }

  private List<NavigationPageItem> createChildItems(final Page parentPage, final ItemCreator itemCreator) {
    List<NavigationPageItem> items = new ArrayList<>();
    Iterator<Page> childPages = parentPage.listChildren(new PageFilter(false, false));
    while (childPages.hasNext()) {
      Page childPage = childPages.next();
      NavigationPageItem item = itemCreator.create(childPage);
      if (item != null) {
        items.add(item);
      }
    }
    return items;
  }

  private List<NavigationPageItem> createChildItemsRecursively(final Page parentPage, final ItemCreator itemCreator,
      final int level, final int maxLevels) {
    return createChildItems(parentPage, new ItemCreator() {
      @Override
      public NavigationPageItem create(final Page pPage) {
        NavigationPageItem item = itemCreator.create(pPage);
        if (item != null) {
          if (level < maxLevels) {
            item.setChildren(createChildItemsRecursively(pPage, itemCreator, level + 1, maxLevels));
          }
        }
        return item;
      }
    });
  }

  private interface ItemCreator {
    NavigationPageItem create(Page pPage);
  }

  private class ValidLinkableItemCreator implements ItemCreator {
    @Override
    public NavigationPageItem create(Page page) {
      NavigationPageItem item = createLinkableItem(page);
      if (item.getLink().isValid()) {
        return item;
      }
      else {
        return null;
      }
    }
  }

}

File: bundles/core/src/main/java/io/wcm/samples/core/business/navigation/impl/NavigationManagerImpl.kt
/*
 * #%L
 * wcm.io
 * %%
 * Copyright (C) 2014 wcm.io
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.wcm.samples.core.business.navigation.impl

import com.day.cq.wcm.api.Page
import com.day.cq.wcm.api.PageFilter
import io.wcm.handler.link.LinkHandler
import io.wcm.handler.url.ui.SiteRoot
import io.wcm.samples.core.business.navigation.NavigationManager
import io.wcm.samples.core.business.navigation.NavigationPageItem
import io.wcm.sling.models.annotations.AemObject
import org.apache.commons.lang3.StringUtils
import org.apache.sling.api.SlingHttpServletRequest
import org.apache.sling.models.annotations.Model
import org.apache.sling.models.annotations.injectorspecific.Self
import java.util.ArrayList

@Model(adaptables = [SlingHttpServletRequest::class], adapters = [NavigationManager::class])
class NavigationManagerImpl : NavigationManager {

  @Self
  private lateinit var siteRoot: SiteRoot

  @Self
  private lateinit var linkHandler: LinkHandler

  @AemObject
  private lateinit var currentPage: Page

  companion object {
    const val FOOTERNAV_RELATIVE_PATH = "/tools/navigation/footernav"
  }

  /**
   * Generation main navigation links.
   * @param maxLevels Max. navigation hierarchy levels
   * @return Root page for navigation
   */
  override fun getMainNavigation(maxLevels: Int): NavigationPageItem {
    val siteRootPage = siteRoot.rootPage ?: currentPage

    val rootItem = createLinkableItem(siteRootPage)
    if (maxLevels > 0) {
      rootItem.children = createChildItemsRecursively(siteRootPage, this::validLinkableItemCreator, 1, maxLevels)
    }
    return rootItem
  }

  /**
   * Generation footer navigation links and navigation structure.
   * If a page '/tools/navigation/footerNav" exists this page is used for the root navigation hierarchy, using the first
   * level of pages as column structure and the next level as link entries.
   * If not present the main navigation sections are used as column structure, and the main navigation pages itself
   * including their children as link entries.
   * @return Root page for navigation
   */
  override fun getFooterNavigation(): NavigationPageItem {
    val footerNavRoot = siteRoot.getRelativePage(FOOTERNAV_RELATIVE_PATH)
    return if (footerNavRoot != null) {
      getFooterNavigationSpecific(footerNavRoot)
    } else {
      val siteRootPage = siteRoot.rootPage ?: currentPage
      getFooterNavigationDerivedFromMainNav(siteRootPage)
    }
  }

  private fun getFooterNavigationSpecific(footerNavRoot: Page): NavigationPageItem {
    return createStructureItem(footerNavRoot).also {
      it.children = createChildItems(footerNavRoot) { childPage ->
        val columnItem = createStructureItem(childPage)
        columnItem.children = createChildItems(childPage, this::validLinkableItemCreator)
        if (columnItem.children.isEmpty()) {
          null
        } else {
          columnItem
        }
      }
    }
  }

  private fun getFooterNavigationDerivedFromMainNav(siteRootPage: Page): NavigationPageItem {
    return createStructureItem(siteRootPage).also {
      it.children = createChildItems(siteRootPage) { childPage ->
        val columnItem = createStructureItem(childPage)
        val linkItems = ArrayList<NavigationPageItem>()

        val mainnavLinkItem = createLinkableItem(childPage)
        if (mainnavLinkItem.link != null && mainnavLinkItem.link.isValid) {
          linkItems.add(mainnavLinkItem)
        }
        linkItems.addAll(createChildItems(childPage, this::validLinkableItemCreator))
        columnItem.children = linkItems
        if (columnItem.children.isEmpty()) {
          null
        } else {
          columnItem
        }
      }
    }
  }

  private fun createStructureItem(page: Page): NavigationPageItem {
    val title = getItemTitle(page)
    return NavigationPageItem(title)
  }

  private fun createLinkableItem(page: Page): NavigationPageItem {
    val title = getItemTitle(page)
    val isCurrentPage = StringUtils.equals(page.path, currentPage.path)
    val link = linkHandler[page].build()
    return NavigationPageItem(title, isCurrentPage, link)
  }

  private fun getItemTitle(page: Page): String {
    return StringUtils.defaultString(page.navigationTitle, page.title)
  }

  private fun createChildItems(parentPage: Page, itemCreator: (Page) -> NavigationPageItem?): List<NavigationPageItem> {
    val items = ArrayList<NavigationPageItem>()
    for (childPage in parentPage.listChildren(PageFilter(false, false))) {
      val item = itemCreator.invoke(childPage)
      if (item != null) {
        items.add(item)
      }
    }
    return items
  }

  private fun createChildItemsRecursively(
      parentPage: Page,
      itemCreator: (Page) -> NavigationPageItem?,
      level: Int,
      maxLevels: Int
  ): List<NavigationPageItem> {
    return createChildItems(parentPage) {
      val item = itemCreator.invoke(it)
      if (item != null && level < maxLevels) {
        item.children = createChildItemsRecursively(it, itemCreator, level + 1, maxLevels)
      }

      item
    }
  }

  private fun validLinkableItemCreator(page: Page): NavigationPageItem? {
    val item: NavigationPageItem = createLinkableItem(page)
    return if (item.link != null && item.link.isValid) {
      item
    } else {
      null
    }
  }
}